whole-tale/dashboard
# Notification widget to receive status updates on "Rebuild"

### Problem
Rebuild needs to notify user when build completes or fails. User needs to be able to view the logs from failed events to fix the broken build.

### Approach
Implement SSE using a custom `EventSource` allowing us to pass the `Girder-Token` header. Also added a log viewer modal that will display pre-formatted text

NOTE: There may still be formatting errors in the log output that we should revisit (e.g. random commas, escape sequences, console codes, etc)

### How to Test
1. Checkout and run this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Launch a Tale, if you haven't already
4. Navigate to the "Run" view and select a Tale
5. Expand the "..." dropdown at the top-center, and choose "Rebuild Tale"
    * After a few seconds, you should see the new `notification-stream` component appear to tell you that the image is building
6. Click on the status (e.g. `BUILDING`) to see the logs
    * You should see that the log output is automatically updated/refreshed every ~2 seconds
    * Once the build is complete, you should receive another notification indicating that your build is successful (e.g. `SUCCESS`)
    * You should be offered on option to "Restart to Use New Image"
7. Modify the `apt.txt` in your workspace to contain a nonsense package name (e.g. `q; isdhadshiasd `)
8. Select "Rebuild Tale" again
    * After a few seconds, you should see the new `notification-stream` component appear to tell you that the image is building
    * After a few more seconds, you should see a new notification appear indicating that the image build failed (e.g. `INVALID`)
9. Click "Mark All as Read" atop the notification-stream
    * Your notification-stream should dismiss itself
    * Subsequent refreshes should **not** bring back dismissed notifications

NOTE: there is still a subtle timeout issue with EventSource's integration with Girder - see https://github.com/Yaffle/EventSource#server-side-requirements
